File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.java
import java.util.*;

class Solution {
    
    final int INF = 987654321;
    
    public int[] solution(int n, int[][] paths, int[] gates, int[] summits) {
        Set<Integer> gateSet = new HashSet<>();
        Set<Integer> summitSet = new HashSet<>();
        
        List<List<Node>> graph = new ArrayList<>();
        for(int i=0;i<=n;i++) {
            graph.add(new ArrayList<>());
        }
        
        for(int[] path : paths) {
            int from = path[0];
            int to = path[1];
            int weight = path[2];
            
            graph.get(from).add(new Node(to, weight));
            graph.get(to).add(new Node(from, weight));
        }
        
        for(int i=0;i<gates.length;i++) {
            gateSet.add(gates[i]);
        }
        
        for(int i=0;i<summits.length;i++) {
            summitSet.add(summits[i]);
        }
        
        int[] dist = new int[n + 1];
        Arrays.fill(dist, INF);
        PriorityQueue<Node> queue = new PriorityQueue<>();
        
        for(int gate : gateSet) {
            queue.offer(new Node(gate, 0));
            dist[gate] = 0;
        }
        
        while(!queue.isEmpty()) {
            Node cur = queue.remove();
            
            if(cur.weight > dist[cur.num]) {
                continue;
            }
            
            if(summitSet.contains(cur.num)) {
                continue;
            }
            
            for(Node next : graph.get(cur.num)) {
                if(gateSet.contains(next.num)) continue;
                int nextWeight = Math.max(cur.weight, next.weight);
                if(dist[next.num] > nextWeight) {
                    dist[next.num] = nextWeight;
                    queue.offer(new Node(next.num, nextWeight));
                }
            }
        }
        
        int[] answer = {-1, INF};
        
        Arrays.sort(summits);
        
        for(int summit : summits) {
            if(answer[1] > dist[summit]) {
                answer[0] = summit;
                answer[1] = dist[summit];
            }
        }
        
        return answer;
    }
}

class Node implements Comparable<Node>{
    int num, weight;
    Node(int num, int weight) {
        this.num = num;
        this.weight = weight;
    }
    
    @Override
    public int compareTo(Node o) {
        return this.weight - o.weight;
    }
}

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.py
from collections import defaultdict
from heapq import heappush, heappop

INF = int(1e9)

def solution(n, paths, gates, summits):
    graph = defaultdict(list)
    
    for u, v, w in paths:
        graph[u].append((v, w))
        graph[v].append((u, w))
    
    summit_set = set(summits)
    gate_set = set(gates)
    
    pq = []
    dist = [INF] * (n + 1)

    for gate in gates:
        heappush(pq, (0, gate))
        dist[gate] = 0
        
    while pq:
        cur = heappop(pq)
        
        if cur[0] > dist[cur[1]]:
            continue
        
        if cur[1] in summit_set:
            continue
        
        for v, w in graph[cur[1]]:
            if v in gate_set:
                continue
                
            vw = max(cur[0], w)
            if dist[v] > vw:
                dist[v] = vw
                heappush(pq, (vw, v))
        
    answer = [[summit, dist[summit]] for summit in summits]
    answer.sort(key=lambda x:(x[1], x[0]))
    
    return answer[0]